src/util/CryptoAmount.ts

```diff
@@ -0,0 +1,102 @@
+import { div, mul } from 'biggystring'
+import { EdgeCurrencyConfig } from 'edge-core-js'
+
+import { RootState } from '../reducers/RootReducer'
+import { getExchangeDenomByCurrencyCode } from '../selectors/DenominationSelectors'
+import { convertCurrency } from '../selectors/WalletSelectors'
+import { asBiggystring } from './cleaners'
+import { mulToPrecision } from './utils'
+
+// TODO: Accept tokenId.
+// Add constructor with EdgeWallet without having to provide pluginId, currencyConfig, etc.
+// Implement fiatValue().
+// Create getter for exchangeAmount for displaying with lower precision.
+
+interface BaseAmountArgs {
+  pluginId: string
+  currencyConfig: EdgeCurrencyConfig
+  currencyCode: string | null
+}
+
+interface ExchangeAmountArgs extends BaseAmountArgs {
+  exchangeAmount: number | string
+}
+
+interface NativeAmountArgs extends BaseAmountArgs {
+  nativeAmount: string
+}
+
+type CryptoAmountConstructorArgs = ExchangeAmountArgs | NativeAmountArgs
+
+/**
+ * One-stop shop for any information pertaining to a crypto asset.
+ * Pass whatever you happen to have, get what you need.
+ *
+ * Usage:
+ * const cryptoAmountWithExchange = CryptoAmount.createWithExchangeAmount({
+ *   pluginId: 'bitcoin',
+ *   currencyConfig,
+ *   exchangeAmount: '100',
+ * })
+ *
+ * const cryptoAmountWithNative = CryptoAmount.createWithNativeAmount({
+ *   pluginId: 'ethereum',
+ *   currencyCode: 'USDC',
+ *   nativeAmount: '12345678',
+ * })
+ */
+export class CryptoAmount {
+  public readonly pluginId: string
+  public readonly chainCode: string
+  public readonly currencyCode: string // Token currency code or chain currency code if no token specified
+  public readonly exchangeAmount: string
+  public readonly nativeAmount: string
```

Comment:
This class makes no separation between *what* value we are talking about, and how we *display* that value. The concepts are jumbled together.

If we want to uniquely identify a value, we need `{ nativeAmount, pluginId, tokenId }`, and that is all. With those three values, we know what coin we are talking about and how much there is. Anything else, we can derive.

Therefore, I would expect the class members to be `tokenId: string, nativeAmountString, currencyConfig: EdgeCurrencyConfig`. The `currencyConfig` object contains the `pluginId`, and is our link to the core.

Then, the class can expose all the existing things using accessor methods, like this:

```ts
get currencyCode(): string {
  const { currencyCode } = this.tokenId == null ? this.currencyConfig.currencyInfo : this.currencyConfig.allTokens[this.tokenId]
  return currencyCode
}

get chainCode(): string {
  return this.currencyConfig.currencyInfo.currencyCode
}

get exchangeAmount(): string {
  const { multiplier } = getExchangeDenom(this.currencyConfig, this.tokenId)
  return div(nativeAmount, multiplier, mulToPrecision(multiplier))
}

get pluginId (): string {
  return this.currencyConfig.currencyInfo.pluginId
}
```

And besides these, we can provide formatting methods like:

```ts
formatDisplayAmount({ redux, precision: ... }): string {
  const {multiplier} = getDisplayDenom(redux, this.currencyConfig, this.tokenId)
  ...
}

formatExchangeAmount({ precision: ... }): string {
  ...
}
```

So you see that the class can be convenient and fully-featured, while still be "lean" with what it stores.

At the end of the day, storing the `currencyCode` as a class member is banned. Absolutely do not store the currency code. We can maybe discuss storing the `exchangeAmount`, but *only* as an optimization.

If you constructor wants to take a `currencyCode`, that's fine. Just mark it `/** @deprecated */` and immediatrely convert it to a `tokenId`, and I have no complaint.

Reply:
If you look at the usage, you see that when we interface with external providers, we often have `exchangeAmount` and `currencyCode` as the input, where we then repeatedly jump through the same hoops to convert values from there. DRYing that out was one of the main design considerations for this class. I can make it more explicit though to make it clear that internally we don't want to start from those inputs.

For the purposes of this PR, I'm not really trying to handle display, I'm mostly keeping the implementation to what I need it for for now, which is reporting `exchangeAmount` and `dollarValues`